### PR TITLE
Media item upload any file and download

### PR DESCRIPTION
**Pull request by dataclouder**
Replaces PR #1123 

Add property `upload_any_file` to `vcd_catalog_media` resource  to enable uploading any file. instead of .ISO files

Adds property `download_to_file` to `vcd_catalog_media` data source, to allow downloading a media item to a file.

### CODE REVIEW

File: .changes/v3.11.0/1123-improvements.md [hunk 1]
```diff
@@ -0,0 +1 @@
+* Add property `upload_any_file` to resource `vcd_catalog_media` to allow uploading any file as catalog media item [GH-1123]
```
Comment (adambarreiro) [inline .changes/v3.11.0/1123-improvements.md:1]: Should we put original PR #1123 or this one? (#1124)

**Review by adambarreiro — APPROVED** [review]
LGTM! Tests passed on VCD 10.5.0

File: .changes/v3.11.0/1123-improvements.md (hunk 1 above)
Comment (dataclouder) [reply .changes/v3.11.0/1123-improvements.md:1]: PR #1123 implemented the resource change. This one (built on top of that branch) implemented the data source change. I'd leave it as it is. If I have to make changes, I'll put both numbers

File: vcd/resource_vcd_catalog_media_test.go [hunk 2]
```diff
@@ -142,32 +152,34 @@ func testAccCheckVcdCatalogMediaExists(mediaName string) resource.TestCheckFunc
...
+			if err != nil {
+				return fmt.Errorf(errorRetrievingOrg, testConfig.VCD.Org+" and error: "+err.Error())
+			}
+
+			catalog, err := adminOrg.GetCatalogByName(testConfig.VCD.Catalog.Name, false)
+			if err != nil {
+				return fmt.Errorf("catalog query %s ended with error: %#v", rs.Primary.ID, err)
+			}
+
+			//mediaName := rs.Primary.Attributes["name"]
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/resource_vcd_catalog_media_test.go]: Yes. Thanks. Removed

**Review by Didainius — APPROVED** [review]
I did some manual play and it worked - thanks

File: website/docs/d/catalog_media.html.markdown [hunk 3]
```diff
@@ -47,6 +47,7 @@ The following arguments are supported:
 * `catalog_id` - (Optional; *v3.8.2+*) The ID of the catalog to which the media file belongs. It's mandatory if `catalog` field is not used.
 * `name` - (Required) Media name in catalog (optional when `filter` is used)
 * `filter` - (Optional; *2.9+*) Retrieves the data source using one or more filter parameters
+* `download_to_file` - (Optional; *3.11+*) Will download the contents of the media item into the given file
```
Comment (lvirbalas) [inline website/docs/d/catalog_media.html.markdown:50]: What are the limitations in size? What about adding a note that there are limitations?

Comment (lvirbalas) [reply website/docs/d/catalog_media.html.markdown:50]: Thanks for the note!

File: vcd/resource_vcd_catalog_media_test.go [hunk 4]
```diff
@@ -142,32 +152,33 @@ func testAccCheckVcdCatalogMediaExists(mediaName string) resource.TestCheckFunc
...
+				return fmt.Errorf(errorRetrievingOrg, testConfig.VCD.Org+" and error: "+err.Error())
+			}
+
+			catalog, err := adminOrg.GetCatalogByName(testConfig.VCD.Catalog.Name, false)
+			if err != nil {
+				return fmt.Errorf("catalog query %s ended with error: %#v", rs.Primary.ID, err)
+			}
+
+			_, err = catalog.GetMediaByName(mediaName, false)
+
```
Comment (lvirbalas) [inline vcd/resource_vcd_catalog_media_test.go:174]: Nit. Extra empty line.

Comment (dataclouder) [reply vcd/resource_vcd_catalog_media_test.go:174]: Removed

File: vcd/resource_vcd_catalog_media_test.go [hunk 5]
```diff
@@ -221,9 +242,14 @@ output "storage_profile_name" {
...
-  org     = "{{.Org}}"
-  catalog = "{{.Catalog}}"
+data "vcd_catalog" "{{.Catalog}}" {
+  org  = "{{.Org}}"
+  name = "{{.Catalog}}"
+}
+
+resource "vcd_catalog_media"  "{{.CatalogMediaName}}" {
+  org  = "{{.Org}}"
+  catalog_id = data.vcd_catalog.{{.Catalog}}.id
```
Comment (adezxc) [inline vcd/resource_vcd_catalog_media_test.go:252]:
```suggestion
  org             = "{{.Org}}"
  catalog_id = data.vcd_catalog.{{.Catalog}}.id
```

File: vcd/resource_vcd_catalog_media_test.go [hunk 6]
```diff
@@ -183,14 +193,25 @@ resource "vcd_catalog_media"  "{{.CatalogMediaName}}" {
   description          = "{{.Description}}"
   media_path           = "{{.MediaPath}}"
   upload_piece_size    = {{.UploadPieceSize}}
-  show_upload_progress = "{{.UploadProgress}}"
+  show_upload_progress = {{.UploadProgress}}
 
   metadata = {
     mediaItem_metadata = "mediaItem Metadata"
     mediaItem_metadata2 = "mediaItem Metadata2"
```
Comment (adezxc) [inline vcd/resource_vcd_catalog_media_test.go:200]:
```suggestion
    mediaItem_metadata   = "mediaItem Metadata"
    mediaItem_metadata2 = "mediaItem Metadata2"
```

**Review by adezxc — APPROVED** [review]
Nice feature, thanks! Only some indentation nits, you can resolve when fixed.

File: vcd/resource_vcd_catalog_media_test.go (hunk 5 above)
Comment (dataclouder) [reply vcd/resource_vcd_catalog_media_test.go:252]: Thanks. Changed

File: vcd/resource_vcd_catalog_media_test.go (hunk 6 above)
Comment (dataclouder) [reply vcd/resource_vcd_catalog_media_test.go:200]: fixed

**Review by lvirbalas — APPROVED** [review]

**Pull request merged**